tests/integration/distribution.go

```diff
@@ -690,197 +696,92 @@ func (s *CCVTestSuite) TestIBCTransferMiddleware() {
-
-			// check remaining consumer rewards allocation
-			expCoinTransferred, expRemaining := tc.rewardsAlloc.TruncateDecimal()
-			s.Require().Equal(expCoinTransferred, coinsTransferred)
-
-			s.Require().Equal(
-				expRemaining,
-				providerKeeper.GetConsumerRewardsAllocation(ctx, chainID).Rewards,
-			)
+	// Iterate over the validators and
+	// store their current voting power and outstanding rewards
+	lastValOutRewards := map[string]sdk.DecCoins{}
+	totalValsRewards := sdk.DecCoins{}
+	for _, val := range s.providerChain.Vals.Validators {
+		valRewards, err := distributionKeeper.GetValidatorOutstandingRewards(s.providerCtx(), sdk.ValAddress(val.Address))
+		s.Require().NoError(err)
+		lastValOutRewards[sdk.ValAddress(val.Address).String()] = valRewards.Rewards
+		totalValsRewards = totalValsRewards.Add(valRewards.Rewards...)
+	}
 
-			// check updated consuemer rewards pool balance
-			newPool := bankKeeper.GetAllBalances(
-				ctx,
-				distributionKeeper.GetDistributionAccount(ctx).GetAddress(),
-			)
+	s.Require().True(totalValsRewards.IsZero())
 
-			s.Require().Equal(newPool.Sub(oldPool...), coinsTransferred)
-		})
+	// store community pool balance
+	lastCommPool := getDistrAcctBalFn(s.providerCtx())
+
+	// execute BeginBlock to trigger the token allocation
+	providerKeeper.BeginBlockRD(s.providerCtx())
+
+	valNum := len(s.providerChain.Vals.Validators)
+	consNum := len(s.consumerBundles)
+
+	// compute the expected validators token allocation by subtracting the community tax
+	rewardsPerConsumerDec := sdk.NewDecCoinsFromCoins(rewardsPerConsumer...)
+	communityTax, err := distributionKeeper.GetCommunityTax(s.providerCtx())
+	s.Require().NoError(err)
+
+	rewardsPerChain, _ := rewardsPerConsumerDec.
+		MulDecTruncate(math.LegacyOneDec().Sub(communityTax)).TruncateDecimal()
+	validatorsExpRewardsPerchain := sdk.NewDecCoinsFromCoins(rewardsPerChain...).QuoDec(math.LegacyNewDec(int64(valNum)))
+	// multiply by the number of consumers
+	validatorsExpRewards := validatorsExpRewardsPerchain.MulDec(math.LegacyNewDec(int64(consNum)))
+
+	// verify the validator tokens allocation
+	// note that the validators have the same voting power to keep things simple
+	for _, val := range s.providerChain.Vals.Validators {
+		valRewards, err := distributionKeeper.GetValidatorOutstandingRewards(s.providerCtx(), sdk.ValAddress(val.Address))
+		s.Require().NoError(err)
+
+		s.Require().Equal(
+			valRewards.Rewards,
+			lastValOutRewards[sdk.ValAddress(val.Address).String()].Add(validatorsExpRewards...),
+		)
 	}
+
+	// check that the total expected rewards is transferred to the distribution module account
+	allocRemPerConsu := providerKeeper.GetConsumerRewardsAllocation(s.providerCtx(), s.consumerChain.ChainID).Rewards
+	totalRewardsDistributed := sdk.NewDecCoinsFromCoins(totalRewards...).Sub(allocRemPerConsu.MulDec(math.LegacyNewDec(int64(consNum))))
+
+	s.Require().Equal(lastCommPool.Add(totalRewardsDistributed...), getDistrAcctBalFn(s.providerCtx()))
```

Comment:
I do not fully understand this assertion. Since we had a `BeginBlockRd`, it seems to me that tokens would have been transferred (?) from the distribution module to the corresponding validators, etc. So, how come the distribution module now contains what the distribution model had plus the `totalRewardsDistributed`? It seems to me we should have a `Sub` instead of an `Add` here (?)

Reply:
After `AllocateTokens` is executed, all rewards are transferred to the distribution module account and the validator's outstanding rewards and the community pool balance are updated. It's only when validators call `WithdrawValidatorCommission` that their token rewards are being transferred to their account.

Reply:
I see. Thank you very much!